testsupport/wait/awaitility.go

```diff
@@ -658,3 +662,125 @@ func (a *Awaitility) listAndReturnContent(resourceKind, namespace string, list c
+			return false, err
+		}
+		for _, obj := range list.Items {
+			obj := obj // required due to memory aliasing until we upgrade to Go 1.22 which fixes this.
+			object, err := w.cast(&obj)
+			if err != nil {
+				return false, fmt.Errorf("failed to cast the object to GVK %v: %w", w.gvk, err)
+			}
+
+			if w.matches(object, predicates) {
+				returnedObject = object
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	return returnedObject, err
+}
+
+// WithNameThat waits for a single object with the provided name in the namespace of the awaitality that additionally
+// matches the provided predicates.
+func (w *Waiter[T]) WithNameThat(name string, predicates ...assertions.Predicate[client.Object]) (T, error) {
+	w.t.Logf("waiting for object of GVK '%s' with name '%s' in namespace '%s' to match additional criteria", w.gvk, name, w.await.Namespace)
+
+	var returnedObject T
+
+	err := wait.Poll(w.await.RetryInterval, w.await.Timeout, func() (done bool, err error) {
+		obj := &unstructured.Unstructured{}
+		obj.SetGroupVersionKind(w.gvk)
+		if err := w.await.Client.Get(context.TODO(), client.ObjectKey{Name: name, Namespace: w.await.Namespace}, obj); err != nil {
+			return false, err
+		}
+		object, err := w.cast(obj)
+		if err != nil {
+			return false, fmt.Errorf("failed to cast the object to GVK %v: %w", w.gvk, err)
+		}
+
+		if w.matches(object, predicates) {
+			returnedObject = object
+			return true, nil
+		}
+
+		return false, nil
+	})
+	return returnedObject, err
+}
+
+func (w *Waiter[T]) cast(obj *unstructured.Unstructured) (T, error) {
+	var empty T
+	raw, err := obj.MarshalJSON()
+	if err != nil {
+		return empty, fmt.Errorf("failed to obtain the raw JSON of the object: %w", err)
+	}
+
+	typed, err := w.await.Client.Scheme().New(w.gvk)
+	if err != nil {
+		return empty, fmt.Errorf("failed to create a new empty object from the scheme: %w", err)
+	}
+
+	err = json.Unmarshal(raw, typed)
+	if err != nil {
+		return empty, fmt.Errorf("failed to unmarshal the raw JSON to the go structure: %w", err)
+	}
+
+	return typed.(T), nil
+}
+
+func (w *Waiter[T]) matches(obj T, predicates []assertions.Predicate[client.Object]) bool {
+	for _, p := range predicates {
+		if !p.Matches(obj) {
+			return false
+		}
+	}
+	return true
+}
+
+// For returns an struct using which one can wait for the objects with the same type as the one provided.
+//
+// Note that this is a recent addition to the utility functions and therefore is not used much. It could
+// be used to cut down the line count of the new utility functions dramatically though.
+func For[T client.Object](t *testing.T, a *Awaitility, obj T) *Waiter[T] {
+	gvks, _, err := a.Client.Scheme().ObjectKinds(obj)
+	require.NoError(t, err, "failed to get the GVK of object %v", obj)
+
+	require.Len(t, gvks, 1, "multiple versions of a single GK not supported but found multiple for object %v", obj)
+
+	return &Waiter[T]{
+		await: a,
+		t:     t,
+		gvk:   gvks[0],
+	}
```

Comment:
The reason is that we don't know which list type to use... I didn't want to include the list type in the func signature, because it would IMHO become too long and weird to use, e.g. `wait.For(t, hostAwaitility, &toolchainv1alpha1.SpaceProvisionerConfig{}, &toolchainv1alpha1.SpaceProvisionerConfigList{}).FirstThat(...)`.